### PR TITLE
DGJ-135 Send data to ODS after Telework form approval/rejection

### PR DESCRIPTION
**Pull request by simensma-fresh**
## Summary
**Sending data to ODS**
Added support for sending Submission data to the ODS using a listener (`SendSubmissionToODSDelegate`), modified the Telework workflow to do so after the submission has been approved/rejected. This just looks up the submission using the `formUrl`, and submits the raw submission with any embedded files replaced with the corresponding filenames.

**Documentation Service**
Added basic config for [Material for MKDocs](https://squidfunk.github.io/mkdocs-material) when running the app locally. This is to encourage us to add documentation as we go, and if we like it, can deploy it to the repos GH Pages. 

How?
- Visit http://localhost:8080/ when running the project locally
- Auto Refreshes on changes
- Any new `md` files should be added to `mkdocs.yml` (`nav` section) so they can be added to the navigation 

## Changes
- Added `org.camunda.bpm.extension.hooks.listeners.SendSubmissionToODSDelegate` delegate for sending submission data to the ODS
- Added / Modified HTTPServiceInvoker / ODSAccessHandler for handling auth + any ODS interactions from within Camunda
- Added material for mkdocs config (mkdocs.yaml, Dockerfile.docs + docker-compose service)

## Screenshots (if applicable)

image
image
image

### CODE REVIEW

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/commons/connector/HTTPServiceInvoker.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    private final Logger LOGGER = Logger.getLogger(HTTPServiceInvoker.class.getName());
+
+    @Autowired
+    private AccessHandlerFactory accessHandlerFactory;
+
+    @Autowired
+    private Properties integrationCredentialProperties;
+
+    @Value("${formsflow.ai.ods.url}")
+    private String odsUrl;
```
Comment (simensma-fresh) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/commons/connector/HTTPServiceInvoker.java:33]: Added this line

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/commons/connector/HTTPServiceInvoker.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    public ResponseEntity<String> execute(String url, HttpMethod method, String payload) {
+            return accessHandlerFactory.getService(getServiceId(url)).exchange(url, method, payload);
+    }
+
+    private String getServiceId(String url) {
+        if(StringUtils.contains(url, getProperties().getProperty("api.url"))) {
+            return "applicationAccessHandler";
+        } else if (StringUtils.contains(url, odsUrl)) {
```
Comment (simensma-fresh) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/commons/connector/HTTPServiceInvoker.java:48]: Added this to delegate the authentication of any ODS requests to the `ODSAccessHandler`

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/SendSubmissionToODSDelegate.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        String endpoint = (String) execution.getVariableLocal("endpoint");
+
+        this.httpServiceInvoker.execute(getEndpointUrl(endpoint), HttpMethod.POST, values);
+    }
+
+    public String getEndpointUrl(String endpoint) {
+        return odsUrl + "/" + endpoint;
+    }
+}
```
Comment (iman-jamali-fw) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/listeners/SendSubmissionToODSDelegate.java:66]: It's great we can use this method to send data to ODS!

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java [hunk 4]
```diff
@@ -161,14 +163,24 @@ private String getSubmissionUrl(String formUrl){
...
             while (dataElements.hasNext()) {
                 Map.Entry<String, JsonNode> entry = dataElements.next();
-                if(StringUtils.endsWithIgnoreCase(entry.getKey(),"_file")) {
+
+                if(!withFileInfo && entry.getValue() != null && entry.getValue().isArray() && entry.getValue().get(0).has("originalName")) {
+                    ArrayNode vals = (ArrayNode) entry.getValue();
+                    String fileNames = StreamSupport.stream(vals.spliterator(), false).map(n -> n.get("originalName").asText())
+                            .collect(Collectors.joining(", "));
+                    fieldValues.put(entry.getKey(), fileNames);
+                } else if(StringUtils.endsWithIgnoreCase(entry.getKey(),"_file")) {
```
Comment (iman-jamali-fw) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java:183]: Man Java is so verbose!

**Review by iman-jamali-fw — APPROVED** [review]
Great work @simensma-fresh! These pieces can be reused for other forms too.

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java [hunk 5]
```diff
@@ -161,14 +163,24 @@ private String getSubmissionUrl(String formUrl){
...
         if(StringUtils.isNotEmpty(submission)) {
             JsonNode dataNode = getObjectMapper().readTree(submission);
             Iterator<Map.Entry<String, JsonNode>> dataElements = dataNode.findPath("data").fields();
             while (dataElements.hasNext()) {
                 Map.Entry<String, JsonNode> entry = dataElements.next();
-                if(StringUtils.endsWithIgnoreCase(entry.getKey(),"_file")) {
+
+                if(!withFileInfo && entry.getValue() != null && entry.getValue().isArray() && entry.getValue().get(0).has("originalName")) {
+                    ArrayNode vals = (ArrayNode) entry.getValue();
+                    String fileNames = StreamSupport.stream(vals.spliterator(), false).map(n -> n.get("originalName").asText())
```
Comment (matbusby-fw) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java:180]: These are the things I was struggling with on my task.  Not having knowledge of some of the classes/utilities available to deal with different types.  Very cool!

**Review by matbusby-fw — APPROVED** [review]
Really awesome work!  The documentation enhancement is a great addition too.

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java (hunk 5 above)
Comment (simensma-fresh) [reply forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java:180]: Yup, Java streams are indeed pretty powerful, but takes some getting used to! https://stackify.com/streams-guide-java-8/ is a good read if interested

File: forms-flow-bpm/Dockerfile [hunk 6]
```diff
@@ -11,11 +11,10 @@ RUN apt-get update \
...
   && rm -rf /var/lib/apt/lists/*
 
-ARG MAVEN_VERSION=3.8.4
+ARG MAVEN_VERSION=3.8.5
 ARG USER_HOME_DIR="/root"
-ARG SHA=a9b2d825eacf2e771ed5d6b0e01398589ac1bfa4171f36154d1b5787879605507802f699da6f7cfc80732a5282fd31b28e4cd6052338cbef0fa1358b48a5e3c8
-ARG BASE_URL=https://apache.osuosl.org/maven/maven-3/${MAVEN_VERSION}/binaries
-
+ARG SHA=89ab8ece99292476447ef6a6800d9842bbb60787b9b8a45c103aa61d2f205a971d8c3ddfb8b03e514455b4173602bd015e82958c0b3ddc1728a57126f773c743
+ARG BASE_URL=https://dlcdn.apache.org/maven/maven-3/${MAVEN_VERSION}/binaries
```
Comment (simensma-fresh) [inline forms-flow-bpm/Dockerfile:17]: The 3.8.4 maven binary disappeared from `apache.osuosl.org` so updated it to the official `dlcdn.apache.org/` repo

File: forms-flow-bpm/Dockerfile [hunk 7]
```diff
@@ -52,4 +51,4 @@ RUN mvn -s /usr/share/maven/ref/settings-docker.xml exec:java -e -Dmain.class=co
 
 EXPOSE 8080
 
-ENTRYPOINT ["/bin/bash", "-c", "lsyncd lsyncd.conf; mvn -s /usr/share/maven/ref/settings-docker.xml package; java -Djava.security.egd=file:/dev/./urandom -jar /tmp/target/forms-flow-bpm.jar"]
\ No newline at end of file
+ENTRYPOINT ["/bin/bash", "-c", "lsyncd lsyncd.conf; mvn -s /usr/share/maven/ref/settings-docker.xml -Dmaven.test.skip=true package; java -Djava.security.egd=file:/dev/./urandom -jar /tmp/target/forms-flow-bpm.jar"]
```
Comment (simensma-fresh) [inline forms-flow-bpm/Dockerfile:54]: Just added in this so we don't have to run the tests every time we make a change to the bpm

File: docker-compose.yml [hunk 8]
```diff
@@ -249,6 +251,15 @@ services:
...
 
+  forms-flow-mailhog:
+    container_name: forms-flow-mailhog
+    hostname: forms-flow-mailhog
+    image: mailhog/mailhog
+    ports:
+      - "1025:1025"
+      - "8025:8025"
+    networks:
+      - forms-flow-network
```
Comment (iman-jamali-fw) [inline docker-compose.yml:264]: Awesome addition! NO more limitations of Mailtrap!

**Pull request merged**